src/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs

```diff
@@ -389,47 +421,215 @@ public static SocketError IoctlInternal(SafeCloseSocket handle, IOControlCode io
                 out optionLength,
                 SafeNativeOverlapped.Zero,
                 IntPtr.Zero);
+            return errorCode == SocketError.SocketError ? GetLastSocketError() : SocketError.Success;
         }
 
         public static unsafe SocketError SetSockOpt(SafeCloseSocket handle, SocketOptionLevel optionLevel, SocketOptionName optionName, int optionValue)
         {
-            return Interop.Winsock.setsockopt(
+            SocketError errorCode = Interop.Winsock.setsockopt(
                 handle,
                 optionLevel,
                 optionName,
                 ref optionValue,
                 sizeof(int));
+            return errorCode == SocketError.SocketError ? GetLastSocketError() : SocketError.Success;
         }
 
         public static SocketError SetSockOpt(SafeCloseSocket handle, SocketOptionLevel optionLevel, SocketOptionName optionName, byte[] optionValue)
         {
-            return Interop.Winsock.setsockopt(
+            SocketError errorCode = Interop.Winsock.setsockopt(
                 handle,
                 optionLevel,
                 optionName,
                 optionValue,
                 optionValue != null ? optionValue.Length : 0);
+            return errorCode == SocketError.SocketError ? GetLastSocketError() : SocketError.Success;
+        }
+
+        public static SocketError SetMulticastOption(SafeCloseSocket handle, SocketOptionName optionName, MulticastOption optionValue)
+        {
+            Interop.Winsock.IPMulticastRequest ipmr = new Interop.Winsock.IPMulticastRequest();
+
+            ipmr.MulticastAddress = unchecked((int)optionValue.Group.GetAddress());
+
+            if (optionValue.LocalAddress != null)
```

Comment:
Missed it initially. I see now that I can scroll to the right by selecting within the code section.